File: cotsSource/org.apache.qpid/dotnet/client-010/client/transport/Struct.cs
/*
*
* Licensed to the Apache Software Foundation (ASF) under one
* or more contributor license agreements.  See the NOTICE file
* distributed with this work for additional information
* regarding copyright ownership.  The ASF licenses this file
* to you under the Apache License, Version 2.0 (the
* "License"); you may not use this file except in compliance
* with the License.  You may obtain a copy of the License at
*
*   http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing,
* software distributed under the License is distributed on an
* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
* KIND, either express or implied.  See the License for the
* specific language governing permissions and limitations
* under the License.
*
*/
using System;
using System.Collections.Generic;
using System.Text;
using org.apache.qpid.transport.codec;

namespace org.apache.qpid.transport
{
	/// <summary> 
	/// Struct
	/// </summary>

    public abstract class Struct : IEncodable
    {
        public  static Struct Create(int type)
        {
            return StructFactory.create(type);
        }

        bool dirty = true;

        public bool Dirty
        {
            get { return dirty; }
            set { dirty = value; }
        }

        public abstract int GetStructType();

        public abstract int GetSizeWidth();

        public abstract int GetPackWidth();

        public int GetEncodedType()
        {
            int type = GetStructType();
            if (type < 0)
            {
                throw new Exception();
            }
            return type;
        }

        private bool IsBit<C, T>(Field<C, T> f)
        {
            return Equals(f.Type, typeof(Boolean));
        }

        private bool Packed()
        {
            return GetPackWidth() > 0;
        }

        private bool Encoded<C, T>(Field<C, T> f)
        {
            return !Packed() || !IsBit(f) && f.Has(this);
        }

        private int GetFlagWidth()
        {
            return (Fields.Count + 7) / 8;
        }

        private int GetFlagCount()
        {
            return 8 * GetPackWidth();
        }

        public abstract void Read(IDecoder dec);

        public abstract void Write(IEncoder enc);

        public abstract Dictionary<String, Object> Fields
        {
            get;
        }

        public override String ToString()
        {
            StringBuilder str = new StringBuilder();
            str.Append(GetType());
            str.Append("(");
            bool first = true;
            foreach (KeyValuePair<String, Object> me in Fields)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                    str.Append(", ");
                }
                str.Append(me.Key);
                str.Append("=");
                str.Append(me.Value);
            }
            str.Append(")");
            return str.ToString();
        }
    }
}

File: cotsSource/org.apache.qpid/java/common/src/main/java/org/apache/qpid/transport/Struct.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.transport;

import java.util.List;
import java.util.Map;

import org.apache.qpid.transport.codec.Decoder;
import org.apache.qpid.transport.codec.Encodable;
import org.apache.qpid.transport.codec.Encoder;


/**
 * Struct
 *
 * @author Rafael H. Schloming
 */

public abstract class Struct implements Encodable
{

    public static Struct create(int type)
    {
        return StructFactory.create(type);
    }

    boolean dirty = true;

    public boolean isDirty()
    {
        return dirty;
    }

    public void setDirty(boolean dirty)
    {
        this.dirty = dirty;
    }

    public abstract int getStructType();

    public abstract int getSizeWidth();

    public abstract int getPackWidth();

    public final int getEncodedType()
    {
        int type = getStructType();
        if (type < 0)
        {
            throw new UnsupportedOperationException();
        }
        return type;
    }

    private final boolean isBit(Field<?,?> f)
    {
        return f.getType().equals(Boolean.class);
    }

    private final boolean packed()
    {
        return getPackWidth() > 0;
    }

    private final boolean encoded(Field<?,?> f)
    {
        return !packed() || !isBit(f) && f.has(this);
    }

    private final int getFlagWidth()
    {
        return (getFields().size() + 7)/8;
    }

    private final int getPaddWidth()
    {
        int pw = getPackWidth() - getFlagWidth();
        assert pw > 0;
        return pw;
    }

    private final int getFlagCount()
    {
        return 8*getPackWidth();
    }

    private final int getReservedFlagCount()
    {
        return getFlagCount() - getFields().size();
    }

    public abstract void read(Decoder dec);

    public abstract void write(Encoder enc);

    public abstract Map<String,Object> getFields();

    public String toString()
    {
        StringBuilder str = new StringBuilder();
        str.append(getClass().getSimpleName());

        str.append("(");
        boolean first = true;
        for (Map.Entry<String,Object> me : getFields().entrySet())
        {
            if (first)
            {
                first = false;
            }
            else
            {
                str.append(", ");
            }
            str.append(me.getKey());
            str.append("=");
            str.append(me.getValue());
        }
        str.append(")");

        return str.toString();
    }

}
